### PR TITLE
Add Content Security Policy Support

### PR DESCRIPTION
**Pull request by pivotal-joe-grandja**
- Adds CSP support to the Java DSL of HeadersConfigurer
- Implementation is pretty straight forward but the documentation is important here so please review and provide feedback on whether it's sufficient

### CODE REVIEW

File: config/src/main/java/org/springframework/security/config/annotation/web/configurers/HeadersConfigurer.java [hunk 1]
```diff
@@ -658,6 +656,79 @@ private HpkpConfig enable() {
...
+		 * @param policyDirectives the security policy directive(s) to write in the response.
+		 *
+		 * @return the {@link ContentSecurityPolicyConfig} for additional configuration
+		 */
+		public ContentSecurityPolicyConfig directives(String policyDirectives) {
+			if (writer == null) {
+				this.writer = new ContentSecurityPolicyHeaderWriter(policyDirectives);
+			} else {
+				// TODO Allow for over-riding previously set directives? Is this really needed?
+				this.writer.setPolicyDirectives(policyDirectives);
```
Comment (rwinch) [inline config/src/main/java/org/springframework/security/config/annotation/web/configurers/HeadersConfigurer.java]: I think this should be removed. User's can always invoke `contentSecurityPolicy` again.

**Comment by rwinch** [comment]
@pivotal-joe-grandja Thanks for the PR! 

I would try to add a link to one or more of the more popular CSP guides (http://www.html5rocks.com/en/tutorials/security/content-security-policy/ or https://developer.mozilla.org/en-US/docs/Web/Security/CSP are nice)

We should also update the [headers section](https://github.com/spring-projects/spring-security/blob/41c6a797c3303b78a6f1ac6d82efeae3773f680c/docs/manual/src/docs/asciidoc/index.adoc#headers) of the reference. 

Also, we will likely merge all of the CSP stuff at once so feel free to update this PR and squash commits as you get the XML support and documentation updated.

**Comment by pivotal-joe-grandja** [comment]
Ok thanks Rob.
I'll apply those updates.

**Comment by rwinch** [comment]
This will fix gh-3763

**Comment by rwinch** [comment]
@pivotal-joe-grandja Thanks for the updates! This is now merged into master via 2f7f2ff5896cbc1e8fbc4e6577c32c3176d6d0c4

**Pull request closed without merging**